src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.WithIndels.gatk4.vcf

```diff
@@ -95,31 +95,31 @@
-20	10067264	.	G	A	2937.06	.	AC=2;AF=1.00;AN=2;DP=73;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.37;QD=32.91;SOR=0.776	GT:AD:DP:GQ:PL	1/1:0,73:73:99:2951,219,0
-20	10067722	.	A	C	2161.06	.	AC=2;AF=1.00;AN=2;DP=59;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=31.31;SOR=0.846	GT:AD:DP:GQ:PL	1/1:0,54:54:99:2175,163,0
+20	10067264	.	G	A	2937.06	.	AC=2;AF=1.00;AN=2;DP=73;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.37;QD=31.31;SOR=0.776	GT:AD:DP:GQ:PL	1/1:0,73:73:99:2951,219,0
+20	10067722	.	A	C	2161.06	.	AC=2;AF=1.00;AN=2;DP=59;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=31.45;SOR=0.846	GT:AD:DP:GQ:PL	1/1:0,54:54:99:2175,163,0
 20	10068172	.	G	A	532.04	.	AC=2;AF=1.00;AN=2;BaseQRankSum=1.453;DP=18;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=56.04;MQRankSum=-0.309;QD=31.30;ReadPosRankSum=-0.436;SOR=0.481	GT:AD:DP:GQ:PL	1/1:1,16:17:25:546,25,0
-20	10068981	.	G	A	2444.06	.	AC=2;AF=1.00;AN=2;DP=63;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=31.45;SOR=0.950	GT:AD:DP:GQ:PL	1/1:0,59:59:99:2458,178,0
-20	10071135	.	C	T	3272.06	.	AC=2;AF=1.00;AN=2;DP=88;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.21;QD=33.27;SOR=1.003	GT:AD:DP:GQ:PL	1/1:0,79:79:99:3286,238,0
-20	10071187	.	G	A	3425.06	.	AC=2;AF=1.00;AN=2;DP=85;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.73;QD=31.26;SOR=1.367	GT:AD:DP:GQ:PL	1/1:0,82:82:99:3439,246,0
+20	10068981	.	G	A	2444.06	.	AC=2;AF=1.00;AN=2;DP=63;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=33.27;SOR=0.950	GT:AD:DP:GQ:PL	1/1:0,59:59:99:2458,178,0
+20	10071135	.	C	T	3272.06	.	AC=2;AF=1.00;AN=2;DP=88;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.21;QD=31.26;SOR=1.003	GT:AD:DP:GQ:PL	1/1:0,79:79:99:3286,238,0
+20	10071187	.	G	A	3425.06	.	AC=2;AF=1.00;AN=2;DP=85;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.73;QD=27.76;SOR=1.367	GT:AD:DP:GQ:PL	1/1:0,82:82:99:3439,246,0
 20	10072505	.	A	G	2114.06	.	AC=2;AF=1.00;AN=2;DP=67;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=60.00;QD=33.03;SOR=0.822	GT:AD:DP:GQ:PL	1/1:0,64:64:99:2128,191,0
-20	10074716	.	G	A	2415.06	.	AC=2;AF=1.00;AN=2;DP=70;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.67;QD=27.76;SOR=1.214	GT:AD:DP:GQ:PL	1/1:0,64:64:99:2429,192,0
-20	10074806	.	G	A	2832.06	.	AC=2;AF=1.00;AN=2;DP=75;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.38;QD=29.09;SOR=0.840	GT:AD:DP:GQ:PL	1/1:0,71:71:99:2846,213,0
-20	10075043	.	T	C	2420.06	.	AC=2;AF=1.00;AN=2;DP=64;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.64;QD=26.08;SOR=1.107	GT:AD:DP:GQ:PL	1/1:0,61:61:99:2434,184,0
-20	10075168	.	C	T	3627.06	.	AC=2;AF=1.00;AN=2;DP=91;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.37;QD=33.85;SOR=0.997	GT:AD:DP:GQ:PL	1/1:0,88:88:99:3641,264,0
+20	10074716	.	G	A	2415.06	.	AC=2;AF=1.00;AN=2;DP=70;ExcessHet=0.0000;FS=0.000;MLEAC=2;MLEAF=1.00;MQ=59.67;QD=29.09;SOR=1.214	GT:AD:DP:GQ:PL	1/1:0,64:64:99:2429,192,0
```

Comment:
I'm seeing some mess in the qual scores and here evidently all the QD scores specifically are broken... are you certain there isn't some ugly gotcha hiding in here? Do you have an explanation for exactly where the divergence from this branch is coming from?

Reply:
Remember the scary place in `AssemblyBasedCallerUtils` where the List of haplotypes to be sorted by their kmer support score had duplicates?  When I fixed the duplicate issue I broke all the tests because there are ties all over the place, so the exact (arbitrary) way in which you might modify the input order while filtering duplicates and sorting makes a difference.

The reason there are so many ties is that the score is extremely non-discriminating: as long as a kmer appears in even one read, it counts equally well, so the scoring often comes out like this: a few haplotypes with insertions leading, since longer haplotypes have more kmers with which to run up the score (I saw your comment about maybe normalzing by haplotype length -- your concern was well-placed), followed by a big tie among a lot of haplotypes with insertions.

I don't take broken tests lightly (well, to be honest I take it more lightly than, say, David Roazen but I do basically care about the integrity of the GATK) so I set some conditional breakpoints, and it turns out that the discrepancies only occur when the cutoff haplotype count occurred in the middle of a tie.  That is, the differences in QUAL and QD only appear when a tie must be broken arbitrarily.

After seeing this I decided to referee ties more permissively and keep everything with a score at least as good as the cutoff.  Thus, in a later commit of this PR, I decided that if the args are set to keep 5 haplotypes but the 4th through 8th have the same score we keep 8, for example.

Reply:
Okay thank you for this in-depth explanation. When I had written this comment out I had not yet internalized that these mismatches ONLY occur on the pileupcaller test files which is considerably lower impact than I had imagined when I saw the number of test files mismatching. That code in particular is less well tested/supported so our threshold for alarm is lower. The primary impact of this part of the codebase is relevant to DRAGEN-GATK which you are about to take ownership of anyway, so I fully expect anything that might be broken here to have a high chance of getting run to ground later in your process. I accept your explanation.... sorry to raise the alarm bells

Reply:
I double-checked this morning by reverting just the haplotype sorting and every DRAGEN test matches master.  It's less elegant because you have to introduce a `List`, then stream it and call `distinct()`, instead of using a `Set` from the beginning.